crates/rust-analyzer/src/diagnostics/to_proto.rs

```diff
@@ -61,21 +61,51 @@ fn location(
     let file_name = resolve_path(config, workspace_root, &span.file_name);
     let uri = url_from_abs_path(&file_name);
 
-    // FIXME: this doesn't handle UTF16 offsets correctly
     let range = lsp_types::Range::new(
-        lsp_types::Position::new(
-            (span.line_start as u32).saturating_sub(1),
-            (span.column_start as u32).saturating_sub(1),
-        ),
-        lsp_types::Position::new(
-            (span.line_end as u32).saturating_sub(1),
-            (span.column_end as u32).saturating_sub(1),
-        ),
+        position(config, span, span.line_start, span.column_start),
+        position(config, span, span.line_end, span.column_end),
     );
 
     lsp_types::Location { uri, range }
 }
 
+fn position(
+    config: &DiagnosticsMapConfig,
+    span: &DiagnosticSpan,
+    line_offset: usize,
+    column_offset: usize,
+) -> lsp_types::Position {
+    let line_index = line_offset - span.line_start;
+
+    let mut true_column_offset = column_offset;
+    if let Some(line) = span.text.get(line_index) {
+        if line.text.chars().count() == line.text.len() {
```

Comment:
It's not public, but we have a `utf16_lines` map which we could use for a fast(er) path.

Reply:
in order to use LineIndex::utf16_lines, we need to build `LineIndex` by snap.file_line_index(file_id). So we need to get file_id first, can you get file_id by file_path ?

Reply:
in `DiagnosticSpan` we only have character offset, we still need to iterate `DiagnosticSpan.text` to get corresponding BytePos of `DiagnosticSpan.column_start` and `DiagnosticSpan.column_end`.

Reply:
i found it( byte_start , byte_end).

Reply:
> we need to build LineIndex by snap.file_line_index(file_id)

That's actually going to be cached.

I don't remember the code very well, but passing in a `LineIndex` might be fine. That should also work inside the tests.

Reply:
I add a new parameter `GlobalStateSnapshot` to `fn position` to build line_index. when encouner error in calling `from_proto::file_id` or `snap.file_line_index`, I will use original solution to get `lsp_types::range`. 

for test, i build a default `GlobalStateSnapshot` to use old test, but in this situation, the computation of `lsp_types::range` will use original solution instead of `to_proto::range(line_index, text_range)`. 

I cannot find any example of initing `GlobalState` in test, do you have any idea to init `GlobalState` in test to test `fn location` with new `to_proto::range(line_index, text_range)`?

Reply:
I like chars().cont() solution more, because it guarantees consistent data -- we are using the text of the file according to rustc's POV. And we shouldn't worry about O(N)/O(1) here I think -- we transfered O(N) text of the span over the wire anyway.